Applications/Spire/Source/Ui/ArrayTableModel.cpp

```diff
@@ -1,38 +1,124 @@
 #include "Spire/Ui/ArrayTableModel.hpp"
+#include <algorithm>
+#include <stdexcept>
 
 using namespace boost;
 using namespace boost::signals2;
 using namespace Spire;
 
-ArrayTableModel::ArrayTableModel() {}
+ArrayTableModel::ArrayTableModel() : m_transaction_level(0) {}
 
-void ArrayTableModel::push(const std::vector<std::any>& row) {}
+void ArrayTableModel::push(const std::vector<std::any>& row) {
+  if(!m_data.empty() && row.size() != get_column_size()) {
+    throw std::out_of_range("row.size() != get_column_size()");
+  }
+  if(row.empty()) {
```

Comment:
I don't think this check is needed.

Reply:
The two-dimensional vector can insert an empty vector. We can imagine the scenario: when the model is empty and an empty vector is added into the model, then the size of the model is 1 and the column of the model is 0 at this time. Then when a non-empty vector is going to be inserted, the column of the model will be not same as the column of the vector.  What should we do? Allow the non-empty vector to be inserted and changed the model's column?

Reply:
In this case line 13 will throw an exception ```row.size() != get_column_size()```.

Reply:
When an empty vector is added the model first, any non-empty vector will not be allowed to be added again.

Reply:
That is mostly correct. The empty row can first be removed and then a non-empty row added. In general all rows must have the same number of columns.

Reply:
So should I add the process of removing a empty row if it exists or keep the check here to prevent an empty row from being added?

Reply:
No, all that's need is to remove lines 15-17.

Reply:
Below is the unit test of insert, can you point out if it meets expectations? 

auto model = ArrayTableModel();
    auto signal_count = 0;
    auto index = 0;
    auto connection = model.connect_operation_signal(
      [&] (const TableModel::Operation& operation) {
        ++signal_count;
        auto add_operation = get<TableModel::AddOperation>(&operation);
        REQUIRE(add_operation != nullptr);
        REQUIRE(add_operation->m_index == index);
      });

    index = 0;
    REQUIRE_NOTHROW(model.insert({}, index));            // here an empty row is insert
    REQUIRE(model.get_row_size() == 1);                        // is the size correct?
    REQUIRE(signal_count == 1);

    index = 0;
    REQUIRE_NOTHROW(model.insert({1, 2, 3}, index));
    REQUIRE(model.get_row_size() == 1);            **// if the empty row above is removed, the row_size is still 1. It's not intuitive that the insert function deletes one row.**
    REQUIRE(signal_count == 2);
    REQUIRE(model.get<int>(0, 0) == 1);
    REQUIRE(model.get<int>(0, 1) == 2);

Reply:
```REQUIRE_NOTHROW(model.insert({1, 2, 3}, index));```

The above line should throw an exception ```std::out_of_range```